ctru-rs/src/services/ndsp/mod.rs

```diff
@@ -0,0 +1,243 @@
+pub mod wave;
+use wave::{WaveInfo, WaveStatus};
+
+use crate::error::ResultCode;
+use crate::services::ServiceReference;
+
+use once_cell::sync::Lazy;
+use std::sync::Mutex;
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum OutputMode {
+    Mono = ctru_sys::NDSP_OUTPUT_MONO,
+    Stereo = ctru_sys::NDSP_OUTPUT_STEREO,
+    Surround = ctru_sys::NDSP_OUTPUT_SURROUND,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum AudioFormat {
+    PCM8Mono = ctru_sys::NDSP_FORMAT_MONO_PCM8,
+    PCM16Mono = ctru_sys::NDSP_FORMAT_MONO_PCM16,
+    ADPCMMono = ctru_sys::NDSP_FORMAT_MONO_ADPCM,
+    PCM8Stereo = ctru_sys::NDSP_FORMAT_STEREO_PCM8,
+    PCM16Stereo = ctru_sys::NDSP_FORMAT_STEREO_PCM16,
+    FrontBypass = ctru_sys::NDSP_FRONT_BYPASS,
+    SurroundPreprocessed = ctru_sys::NDSP_3D_SURROUND_PREPROCESSED,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum InterpolationType {
+    Polyphase = ctru_sys::NDSP_INTERP_POLYPHASE,
+    Linear = ctru_sys::NDSP_INTERP_LINEAR,
+    None = ctru_sys::NDSP_INTERP_NONE,
+}
+
+pub struct Channel {
+    id: i32,
+}
+
+static NDSP_ACTIVE: Lazy<Mutex<usize>> = Lazy::new(|| Mutex::new(0));
+
+#[non_exhaustive]
+pub struct Ndsp {
+    _service_handler: ServiceReference,
+}
+
+impl Ndsp {
+    pub fn init() -> crate::Result<Self> {
+        let _service_handler = ServiceReference::new(
+            &NDSP_ACTIVE,
+            false,
+            || {
+                ResultCode(unsafe { ctru_sys::ndspInit() })?;
+
+                Ok(())
+            },
+            || unsafe {
+                ctru_sys::ndspExit();
+            },
+        )?;
+
+        Ok(Self { _service_handler })
+    }
+
+    /// Return the representation of the specified channel.
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if the channel id is not between 0 and 23.
+    pub fn channel(&self, id: u8) -> crate::Result<Channel> {
+        if id > 23 {
+            return Err(crate::Error::InvalidChannel(id.into()));
+        }
+
+        Ok(Channel { id: id.into() })
+    }
```

Comment:
Maybe this can be moved to `Channel::new`? It doesn't seem require a NDSP service reference, unless channel operations will fail until `ndspInit` is called?.

Reply:
They won’t fail, but they also won’t do anything either. Since `Channel` functionality is entirely dependent on NDSP, i’d keep it this way.